### PR TITLE
use taskQueueId internally to the queue (tasks)

### LINKED ISSUE #3578: use taskQueueId internally to the queue
This bug covers refactoring the queue to use taskQueueId internally, converting back and forth to provisionerId/workerType as necessary to keep external interfaces unchanged.

### PR DESCRIPTION
**Pull request by alopezz**
Github Bug/Issue: Fixes #3578 (part 3)

Making this change in the tasks table with the usual migration mechanism would probably cause outage problems due to the typical size of this table in known deployments.

#3749 enables the implementation of online migrations, providing a way to apply these changes gradually. This PR will implement the online migration and downgrade needed for this, ~and make the appropriate changes in the queue service to use the new naming~.

### CODE REVIEW

**Comment by djmitche** [comment]
I'm writing a blog post about DB migrations now.  Is it OK if I link here and mention you in the blog post?

**Comment by alopezz** [comment]
Sure, go ahead!

**Comment by djmitche** [comment]
http://code.v.igoro.us/posts/2020/12/tc-db-part-3.html - thanks!

**Comment by alopezz** [comment]
> http://code.v.igoro.us/posts/2020/12/tc-db-part-3.html - thanks!

Thank _you_ for the mention! :smile: And the series of posts is quite helpful to get some more context. Now I better see this through!

**Comment by alopezz** [comment]
@djmitche as per what I told you on matrix, I've pushed the current state of what I was working with when I reported being stumped.

The scripts are very close to the example in the readme, since it matches what we need here.

This fails all tests due to `AssertionError [ERR_ASSERTION]: did not see multiple batches of online downgrade`. The posgres output shows this, which I'm guessing is the culprit:

```
2020-12-04 06:28:21.531 UTC [328] ERROR:  column "worker_type" does not exist at character 27
2020-12-04 06:28:21.531 UTC [328] QUERY:  SELECT * from tasks where worker_type is null or provisioner_id is null limit 1
2020-12-04 06:28:21.531 UTC [328] CONTEXT:  PL/pgSQL function online_downgrade_v56_is_complete() line 3 at PERFORM
2020-12-04 06:28:21.531 UTC [328] STATEMENT:  select * from online_downgrade_v56_is_complete()
```

My assessment of why this happens:

Looking into the code and trying to figure out what was going on, I saw that the ordering of the downgrades is set here:  https://github.com/taskcluster/taskcluster/blob/50b3ebb08ecfd6708f14c9cd907a57e08666147c/libraries/postgres/src/Database.js#L242-L251

That is, the online downgrade is done before the regular download. IIUC, this means that we first run the relevant `online_downgrade_*` functions until `online_downgrade_v56_is_complete()` returns true, and then we run the `downgradeScript` written in the version's yml file.

However, the _old_ columns are added back in the `downgradeScript`:
https://github.com/taskcluster/taskcluster/blob/2e9783f84823c092912b0bf1afdac88d3d26673d/db/versions/0056.yml#L46

That would mean that we shouldn't be able to repopulate the columns with this order.

**Comment by alopezz** [comment]
I made a few changes in the direction I mentioned, mainly changing the order so that `runDowngrade` gets run before `runOnlineDowngrade`; and I got this to work, at least in terms of passing the tests I've written so far. I get the feeling this may still fail, though, if `get_task` is called during the online downgrade for rows that may have not yet had their old fields repopulated (haven't had the time to test this yet though). But I'm struggling a bit to wrap my head around how the downgrade procedure should work without running into any pitfalls.

Now if this is the way it's supposed to look, I also think I'd rather define the online migration and online downgrade functions in the same file, so that I can avoid the `if exists` and `or replace` when dropping and creating them.

There's another point I found. According to the [examples in the readme file](https://github.com/taskcluster/taskcluster/tree/main/db#migrations-and-downgrades), I should be using: https://github.com/taskcluster/taskcluster/blob/0dcb9e9af3c4054576a3eecab53a5965a50679e4/db/versions/0056.yml#L6

instead of the usual way, to avoid a full scan of the table which may slow down the migration. The problem I've found with this is that while it may be functionally equivalent, it's not exactly the same, and `yarn generate` complains if I try to add `not null` to the line: https://github.com/taskcluster/taskcluster/blob/0dcb9e9af3c4054576a3eecab53a5965a50679e4/db/tables.yml#L174

I don't know if this is supposed to be acceptable, but I thought it may at least be worth mentioning.

**Comment by alopezz** [comment]
>  I get the feeling this may still fail, though, if `get_task` is called during the online downgrade for rows that may have not yet had their old fields repopulated (haven't had the time to test this yet though). 

I've tested both `get_task` and `get_tasks_by_task_group` in a way that they would fail during online downgrade. Then, what I did to make those functions work during online downgrade was to redefine them on the same version as the online migration so that they try to get the value for provisioner_id and worker_type from the column but fall back to getting it from splitting the task_queue_id.

So the current implementation of the online migration/downgrade would seem to work. I probably still want to refine the tests even though I think they are thorough enough. Before I move on, I want some feedback on the issues raised in my comments above, especially regarding those changes that fall outside of the SQL migration scripts.

Also, if it's OK, I think I may prefer to split this into two PR's, and limit the scope of this one to the migration and DB changes to get it out of the way. In that case, I would not deprecate the old functions nor leave the new `*_tqid` ones implemented, and I would do that in a second PR where I would also change the code in the queue service.

**Review by djmitche — CHANGES_REQUESTED** [review]
> I don't know if this is supposed to be acceptable, but I thought it may at least be worth mentioning.

Maybe someday we could "translate" that constraint to "not null" in `tables.sql`, but for the moment I think it's fine.  I am surprised it's not showing up in `schema.md`, though.  #4063.

I agree that dropping the functions and using `if not exists` is awkward.  Failing to do that might leave "stale" migration functions around that would be run unexpectedly.  I suspect that the JS that's running these migrations should just remove the expected functions (if they exist) before calling the migrationScript and downgradeScript - then there's no need to worry about dropping them at all, nor of executing "stale" functions.  I'm happy to add that in a followup if you'd like - just file an issue.

This patch looks fantastic.  I agree that the tests are sufficiently thorough.  Maybe a good double-check would be to introduce bugs in the various redefined DB functions and check that the tests fail?

Good spot that as of the completion of version 0055's migrationScript, the table invariant is that provisioner_id/worker_type OR task_queue_id are set, or maybe both.  That means the redefined functions in 0055 must account for either of those possibilities -- as you've done.  That's definitely a subtle point, but I'm not sure how to make it less subtle..

One thing that's missing is to use the `state` to start the migration batch at the next taskId that needs attention.  As written, toward the end of the process the batch will spend most of its time scanning rows where the columns are already not null.

The other thing to change is to define the online downgrade functions in the `downgradeScript` parameter in version 0056, rather than `migrationScript`.  With that change, I think it makes a little more sense to have 0056's online downgrade be the "mirror image" of 0055's online migration, as it's more of a "followup to 0056's downgradeScript" mirroring the "followup to 0055's migrationScript".

Leaving the function changes to a second (third!) PR sounds fine.

File: db/test/versions/0056_test.js
Comment (alopezz) [reply db/test/versions/0056_test.js:95]: I'm not sure I understand what you are trying to say here. IIUC, `concurrentCheck` is indeed run before the downgradeScript, due to:
https://github.com/taskcluster/taskcluster/blob/06b8661b2741eee74d0c88912cb0920a35e7acf8/db/test/helper.js#L310-L316
and
https://github.com/taskcluster/taskcluster/blob/06b8661b2741eee74d0c88912cb0920a35e7acf8/db/test/helper.js#L354-L355

And that is why the try-catch is needed here, to only try to check `get_task` when those columns have already been created. The conditional here is only to ensure that we only omit the checks due to missing columns and not due to any other exception that it may occur (even if that's unlikely).

Comment (djmitche) [reply db/test/versions/0056_test.js:95]: You are absolutely right.  I tested this by deleting the conditional, when I should have replaced it with `throw err`.  Sorry!

**Comment by alopezz** [comment]
> I agree that dropping the functions and using `if not exists` is awkward. Failing to do that might leave "stale" migration functions around that would be run unexpectedly. I suspect that the JS that's running these migrations should just remove the expected functions (if they exist) before calling the migrationScript and downgradeScript - then there's no need to worry about dropping them at all, nor of executing "stale" functions. I'm happy to add that in a followup if you'd like - just file an issue.

I wouldn't mind trying to add this myself in this PR, given that I've already had to touch code in the postgres library to make things work; and that it could be tested with this online migration.

> One thing that's missing is to use the `state` to start the migration batch at the next taskId that needs attention. As written, toward the end of the process the batch will spend most of its time scanning rows where the columns are already not null.

:+1: That's on my checklist. Will do this optimization now that I have this basic working version.

> The other thing to change is to define the online downgrade functions in the `downgradeScript` parameter in version 0056, rather than `migrationScript`. With that change, I think it makes a little more sense to have 0056's online downgrade be the "mirror image" of 0055's online migration, as it's more of a "followup to 0056's downgradeScript" mirroring the "followup to 0055's migrationScript".

Sounds good :+1: 

Probably I'd also like to update the relevant documentation with what we're learning from this and to mention this as an example.

**Comment by djmitche** [comment]
> Probably I'd also like to update the relevant documentation with what we're learning from this and to mention this as an example.

That'd be *fantastic* ❤️

**Comment by alopezz** [comment]
I've addressed most of the points, I think. @djmitche Before updating the documentation, I want to ask again if it's OK to implement the dropping of the migration functions in the JS code to avoid having to worry about that in the migration scripts. To do that, I would like to understand better what we are supposed to achieve or prevent by dropping them, so that I know exactly at which point of the sequence they need to be dropped. Actually, I'm still not sure if I'm currently dropping the downgrade functions in the current scripts.

**Comment by djmitche** [comment]
I think that's OK to implement here.  The idea would be to drop them just before calling the scripts that might redefine them, so that the test for the functions' existence after running that script accurately reflect whether or not the script created the functions.  That was a lot of words, sorry :)

The big picture I see is that the online-migration functions are just an extension of the `migrationScript`: "once this transaction is done, here's some extra work to complete before considering this migration truly finished".  Similar, online downgrades are an extension of `downgradeScript`.  So the online migration/downgrade functions should be defined in `migrationScript`/`downgradeScript`, and the JS should drop any such functions before running those scripts so that it can accurately tell if the scripts create the functions.

The bug this is preventing is unlikely: some previous run left a `_batch` function around that's no longer relevant, but the JS sees it and runs it anyway.  I don't mind leaving unused functions in the DB, but it'd be fine to *also* drop the functions once they have been executed fully.

Maybe this is a lot of talk about a very minor detail -- it's fine to leave this to a followup, too.

**Comment by alopezz** [comment]
Actually that was exactly the kind of response I was hoping for, so thank you :smiley: .

It is relatively easy to implement this, and as I said, the context of this PR looked like a good opportunity for this kind of refinement. So I've implemented it (dropping before and after the complete process); you can take a look and tell me if you think the approach looks right.

This is ready for review implementation-wise. I'll update the docs next; and I also see that I will have to bump the db version numbers again :sweat_smile:.

**Comment by djmitche** [comment]
I had a peek at the latest commit and it looks perfect 👍🏽 

Sorry about the DB versions!  `infrastructure/renumber-db-version.js` might be useful..

**Comment by djmitche** [comment]
(I merged from main again..)

**Comment by alopezz** [comment]
> (I merged from main again..)

You missed renaming the online db functions to match version numbers, probably that's why tests are failing. I also referred to these versions in the docs, so that would also have to be updated before the final merge. (I'm AFK for the rest of the day today)

**Comment by djmitche** [comment]
I'll sort it out :)

**Comment by djmitche** [comment]
In addition to renumbering, I fixed a few of my goofs in subsequent commits.  Take a look and, if you're happy and the tests are green, we can get this merged!  Yay!

File: libraries/postgres/src/Database.js
Comment (alopezz) [inline libraries/postgres/src/Database.js:261]: Much better! :smiley:

File: libraries/postgres/src/Database.js
Comment (alopezz) [inline libraries/postgres/src/Database.js:170]: :+1: I missed these when adding auto-dropping!

**Comment by alopezz** [comment]
> In addition to renumbering, I fixed a few of my goofs in subsequent commits. Take a look and, if you're happy and the tests are green, we can get this merged! Yay!

LGTM! I only had to fix a nit in the doc (there was still a sentence that said that the version number in downgrade functions had to refer to the previous version). Ready to merge as far as I'm concerned!

**Review by djmitche — APPROVED** [review]

**Comment by djmitche** [comment]
Thanks @alopezz -- that was an incredibly complex PR!

**Pull request merged**
